src/crimson/os/seastore/transaction_manager.h

```diff
@@ -292,66 +292,213 @@ class TransactionManager : public ExtentCallbackInterface {
-      existing_paddr,
-      ext.get()
-    ).si_then([ext=std::move(ext), laddr_hint, this](auto &&ref) {
-      ceph_assert(laddr_hint == ref->get_key());
-      return epm->read(
-        ext->get_paddr(),
-	ext->get_length(),
-	ext->get_bptr()
-      ).safe_then([ext=std::move(ext)] {
-	return map_existing_extent_iertr::make_ready_future<TCachedExtentRef<T>>
-	  (std::move(ext));
+    LBAMappingRef &&pin,
+    extent_len_t remap_offset,
+    extent_len_t remap_len) {
+    LOG_PREFIX(TransactionManager::remap_pin);
+    auto remap_laddr = pin->get_key() + remap_offset;
+    auto remap_paddr = pin->get_val().add_offset(remap_offset);
+    // FIXME: paddr can be absolute and pending
+    ceph_assert(pin->get_val().is_absolute());
+    ceph_assert(remap_offset < pin->get_length());
+    ceph_assert(remap_offset + remap_len <= pin->get_length());
+    ceph_assert(remap_len != 0);
+    ceph_assert(remap_offset % cache->get_block_size() == 0);
+    ceph_assert(remap_len % cache->get_block_size() == 0);
+
+    SUBDEBUGT(seastore_tm,
+      "original laddr: {}, original paddr: {}, original length: {}"
+      " remap laddr: {}, remap paddr: {}, remap length: {}", t,
+      pin->get_key(), pin->get_val(), pin->get_length(),
+      remap_laddr, remap_paddr, remap_len);
+    return cache->get_extent_if_cached(
+      t, pin->get_val(), T::TYPE
+    ).si_then([this, &t, original_laddr = pin->get_key(),
+              original_len = pin->get_length(),
+              remap_laddr, remap_paddr, remap_len](auto ext) {
+      std::optional<ceph::bufferptr> original_bptr;
+      if (ext) {
+        ceph_assert(ext->is_clean()); // should not remap dirty extents
+        ceph_assert(ext->get_length() == original_len);
+        original_bptr = ext->get_bptr();
+      }
+      return alloc_remapped_extent_with_retire<T>(
+        t,
+        remap_laddr,
+        remap_paddr,
+        remap_len,
+        original_laddr,
+        std::move(original_bptr)
+      ).handle_error_interruptible(
+          remap_pin_iertr::pass_further{},
+          crimson::ct_error::assert_all{
+            "TransactionManager::remap_pin hit invalid error"
+          }
+      ).si_then([remap_laddr](auto &&rpin) {
+        ceph_assert(rpin->get_key() == remap_laddr);
+        return remap_pin_iertr::make_ready_future<LBAMappingRef>(
+          std::move(rpin));
       });
     });
   }
 
+  /**
+   * overwrite_pin
+   *
+   * Overwrite extent inside the original extent.
+   * Return the tuple of splited extents and overwrite extent.
+   * One of left pin or right pin can be nullptr.
+   */
+  using overwrite_pin_iertr = base_iertr;
+  template <typename T>
+  using overwrite_pin_ret = overwrite_pin_iertr::future<
+    std::tuple<LBAMappingRef, TCachedExtentRef<T>, LBAMappingRef>>;
+  template <typename T>
+  overwrite_pin_ret<T> overwrite_pin(
+    Transaction &t,
+    LBAMappingRef &&pin,
+    extent_len_t overwrite_offset,
+    extent_len_t overwrite_len,
+    ceph::bufferlist bl) {
```

Comment:
cc @xxhdx1985126 

It seems more generic to merge `remap_pin()` and `overwrite_pin()` into a single TM interface when need to remap a pin into one or more smaller new pins:
```c++
struct remap_entry {
  extent_len_t offset;
  extent_len_t len;
};
template <typename T>
LBAMappingRef
remap_pin_iertr::future<std::vector<LBAMappingRef>>
remap_pin(
  Transaction &t,
  LBAMappingRef pin,
  const std::vector<remap_entry> &remaps);
```

Another reason is that it is not TM's responsiblity to deal with `bl` for `ObjectDataBlock`.

Reply:
Yeah, maybe like the `LBAManager::split_mapping()` interface in the seastore clone code: https://github.com/xxhdx1985126/ceph/commit/7828f304ede871824608e3f82d573273a72d8b63

Reply:
Looks the proposed new interface can cover all the requirements from remap_pin(), overwrite_pin(), and split_extent().

Reply:
I am working on this